src/libxively/sft/xi_sft_logic.c

```diff
@@ -131,6 +144,89 @@ xi_sft_send_file_status( const xi_sft_context_t* context,
     ( *context->fn_send_message )( context->send_message_user_data, message_file_status );
 }
 
+xi_state_t xi_sft_select_next_resource_to_download( xi_sft_context_t* context )
+{
+    int32_t selected_index       = -1;
+    context->update_current_file = NULL;
+    uint16_t i                   = 0;
+    for ( ; i < context->update_message_fua->file_update_available.list_len; ++i )
+    {
+        if ( 0 <= context->updateable_files_download_order[i] )
+        {
+            selected_index = context->updateable_files_download_order[i];
+            context->updateable_files_download_order[i] = -1;
+            break;
+        }
+    }
+
+    if ( selected_index >= context->update_message_fua->file_update_available.list_len )
+    {
+        return XI_ELEMENT_NOT_FOUND;
+    }
+
+    if ( -1 != selected_index )
+    {
+        printf( "Selected index: %d\n", selected_index );
+        context->update_current_file =
+            &( context->update_message_fua->file_update_available.list[selected_index] );
+    }
+
+    return XI_STATE_OK;
+}
+
+xi_state_t xi_sft_order_resource_downloads( xi_sft_context_t* context )
+{
+    xi_state_t state = XI_STATE_OK;
+    uint16_t i       = 0; /* forward declaration to suppress toolchain warnings */
+
+    const uint16_t num_incoming_resources =
+        context->update_message_fua->file_update_available.list_len;
+
+    context->updateable_files_download_order =
```

Comment:
from line 185-192 this could be solved with the internal allocators:
```

```

Reply:
I started with these, but the compiler produces warnings (and errors) that the buffers aren't always used due to a potential if path.   I'm not sure I understand why.

Reply:
for instance, for the first one where we aqllocate the download order array:

/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/sft/xi_sft_logic.c:173:5: error: variable 'resource_names' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
    XI_ALLOC_BUFFER_AT( int32_t, context->updateable_files_download_order,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:83:5: note: expanded from macro 'XI_ALLOC_BUFFER_AT'
    XI_CHECK_MEMORY( out, state );                                                       \
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:65:33: note: expanded from macro 'XI_CHECK_MEMORY'
#define XI_CHECK_MEMORY( a, s ) XI_CHECK_CND( ( a ) == 0, XI_OUT_OF_MEMORY, s )
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:30:10: note: expanded from macro 'XI_CHECK_CND'
    if ( ( cnd ) )                                                                       \
         ^~~~~~~
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/sft/xi_sft_logic.c:215:19: note: uninitialized use occurs here
    XI_SAFE_FREE( resource_names );
                  ^~~~~~~~~~~~~~
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:59:12: note: expanded from macro 'XI_SAFE_FREE'
    if ( ( a ) )                                                                         \
           ^
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/sft/xi_sft_logic.c:173:5: note: remove the 'if' if its condition is always false
    XI_ALLOC_BUFFER_AT( int32_t, context->updateable_files_download_order,
    ^
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:83:5: note: expanded from macro 'XI_ALLOC_BUFFER_AT'
    XI_CHECK_MEMORY( out, state );                                                       \
    ^
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:65:33: note: expanded from macro 'XI_CHECK_MEMORY'
#define XI_CHECK_MEMORY( a, s ) XI_CHECK_CND( ( a ) == 0, XI_OUT_OF_MEMORY, s )
                                ^
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/xi_macros.h:30:5: note: expanded from macro 'XI_CHECK_CND'
    if ( ( cnd ) )                                                                       \
    ^
/Users/ddellabitta/dev/xively/xively-client-c/src/libxively/sft/xi_sft_logic.c:176:9: note: variable 'resource_names' is declared here
        char** resource_names = xi_alloc( num_incoming_resources * sizeof( char* ) );
        ^
1 error generated.

Reply:
true, then use this:

```
char** resource_names = NULL;

    XI_ALLOC_BUFFER_AT( int32_t, context->updateable_files_download_order,
                        sizeof( int32_t ) * num_incoming_resources, state );
    XI_ALLOC_BUFFER_AT( char*, resource_names, sizeof( char* ) * num_incoming_resources,
                        state );
```

Reply:
or the order change solves the issue too:
```
    XI_ALLOC_BUFFER( char*, resource_names, sizeof( char* ) * num_incoming_resources,
                     state );
    XI_ALLOC_BUFFER_AT( int32_t, context->updateable_files_download_order,
                        sizeof( int32_t ) * num_incoming_resources, state );
```

Reply:
blort, nice. Thanks!
:+1: